source/common/network/socket_option_impl.cc

```diff
@@ -0,0 +1,71 @@
+#include "common/network/socket_option_impl.h"
+
+#include "envoy/common/exception.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+bool SocketOptionImpl::setOption(Socket& socket, Socket::SocketState state) const {
+  if (state == Socket::SocketState::PreBind) {
```

Comment:
Please also move the `SOL_SOCKET, SO_REUSEADDR` option from `TcpListenSocket::TcpListenSocket`, so that all socket options are set in this one library.
This will require either defining a new `setTcpSocketOption` method or passing `setsockopt`'s `level` as a parameter to use `SOL_SOCKET`.

Reply:
I will leave a TODO, I think this is orthogonal and that setsockopt doesn't cause any #ifdef tangle. So, in the interest of unblocking #2719 will skip for now.